pkg/volume/csi/csi_client.go

```diff
@@ -949,3 +971,27 @@ func (c *csiDriverClient) nodeGetVolumeStatsV1(
 	}
 	return metrics, nil
 }
+
+func isFinalError(err error) bool {
+	// Sources:
+	// https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+	// https://github.com/container-storage-interface/spec/blob/master/spec.md
+	st, ok := status.FromError(err)
+	if !ok {
+		// This is not gRPC error. The operation must have failed before gRPC
+		// method was called, otherwise we would get gRPC error.
+		// We don't know if any previous volume operation is in progress, be on the safe side.
+		return false
+	}
+	switch st.Code() {
+	case codes.Canceled, // gRPC: Client Application cancelled the request
```

Comment:
How about some code like PERMISSION_DENIED, UNAUTHENTICATED?

Reply:
Elsewhere we have assumed those error codes to be "final" and unless CSI driver deployment is fixed, this error will keep happening.

Reply:
Since user is not authenticated or authorized we can safely assume that the operation did not even start. If we mark a volume as uncertain, we would most probably get PERMISSION_DENIED or UNAUTHENTICATED error during Unpublish / Unstage.